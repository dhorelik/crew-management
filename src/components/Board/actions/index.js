import { UPDATE_STATUS } from '../constants/actionTypes'


export const updateStatus = (id, newStatus, oldStatus) => ({
    type: UPDATE_STATUS,
    id,
    newStatus,
    oldStatus
})
