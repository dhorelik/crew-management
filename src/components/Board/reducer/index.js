import initialState from './initialState'
import { UPDATE_STATUS } from '../constants/actionTypes'


export default (state = initialState, action) => {
    switch (action.type) {
        case UPDATE_STATUS:
            // TODO shouldn't handle data transformation in reducer, move to action

            return {
                ...state,
                [action.oldStatus]: state[action.oldStatus].filter(item => item !== action.id),
                [action.newStatus]: state[action.newStatus].concat(action.id)
            }

        default:
            return state
    }
}
