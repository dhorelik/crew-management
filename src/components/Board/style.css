.board {
  display: flex;
  flex-direction: row; }
  .board__column {
    flex: 1; }
