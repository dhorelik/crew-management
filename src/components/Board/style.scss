.board {
    display: flex;
    flex-direction: row;

    &__column {
        flex: 1;
    }
}
