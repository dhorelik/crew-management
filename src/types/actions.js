export type UpdateStatusAction = {
    type: string,
    id: string,
    newStatus: string,
    oldStatus: string
}

export type Action = UpdateStatusAction
