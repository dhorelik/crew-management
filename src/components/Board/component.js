import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Card from 'components/Card'
import './style.css'


export default class Board extends Component {
    static propTypes = {
        candidates: PropTypes.shape({
            byId: PropTypes.shape,
            applied: PropTypes.arrayOf(PropTypes.string),
            interviewing: PropTypes.arrayOf(PropTypes.string),
            hired: PropTypes.arrayOf(PropTypes.string),
        }).isRequired,
        initializeBoard: PropTypes.func.isRequired,
        updateStatus: PropTypes.func.isRequired
    }

    componentWillMount() {
        this.props.initializeBoard()
    }

    handleStatusUpdate = (id, newStatus, oldStatus) => {
        this.props.updateStatus(id, newStatus, oldStatus)
    }

    render() {
        return (
            <div className='board'>
                <div className='board__column'>
                    <h3 className='board__title'>Applied</h3>

                    { this.props.candidates.applied.map(candidateId => <Card
                        key={candidateId}
                        handleRight={() => this.handleStatusUpdate(candidateId, 'interviewing', 'applied')}
                        {...this.props.candidates.byId[candidateId]}/>)
                    }
                </div>

                <div className='board__column'>
                    <h3 className='board__title'>Interviewing</h3>

                    { this.props.candidates.interviewing.map(candidateId => <Card
                        key={candidateId}
                        handleLeft={() => this.handleStatusUpdate(candidateId, 'applied', 'interviewing')}
                        handleRight={() => this.handleStatusUpdate(candidateId, 'hired', 'interviewing')}
                        {...this.props.candidates.byId[candidateId]}/>)
                    }
                </div>

                <div className='board__column'>
                    <h3 className='board__title'>Hired</h3>

                    { this.props.candidates.hired.map(candidateId => <Card
                        key={candidateId}
                        handleLeft={() => this.handleStatusUpdate(candidateId, 'interviewing', 'hired')}
                        {...this.props.candidates.byId[candidateId]}/>)
                    }
                </div>
            </div>
        )
    }
}
