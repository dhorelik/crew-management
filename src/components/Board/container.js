import { connect } from 'react-redux'
import { initializeBoard } from './actions'
import component from './component'


export default connect(
    null,
    { initializeBoard }
)(component)
