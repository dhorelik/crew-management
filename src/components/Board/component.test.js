import React from 'react'
import renderer from 'react-test-renderer'
import Board from './component'
import { updateStatus } from './actions'


// TODO sample data should be kept in constants elsewhere
const mockData = {
    byId: {
        'NL 82 76 72 J': {
            name: {
                title: 'ms',
                first: 'katherine',
                last: 'lane'
            },
            picture: {
                large: 'https://randomuser.me/api/portraits/women/40.jpg'
            }
        }
    },
    applied: [ 'NL 82 76 72 J' ],
    interviewing: [],
    hired: []
}

describe('Board', () => {
    it('should render consistently', () => {
        const component = renderer.create(<Board candidates={mockData} updateStatus={updateStatus} />)
        expect(component.toJSON()).toMatchSnapshot()
    })
})
