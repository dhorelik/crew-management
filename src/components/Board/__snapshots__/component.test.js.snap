// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render consistently 1`] = `
<div
  className="board"
>
  <div
    className="board__column"
  >
    <h3
      className="board__title"
    >
      Applied
    </h3>
    <div
      className="card"
    >
      <img
        alt="katherine lane"
        className="card__image"
        src="https://randomuser.me/api/portraits/women/40.jpg"
      />
      <p
        className="card__name"
      >
        katherine
         
        lane
      </p>
      <a
        className="card__link card__link_position_right"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex="0"
      >
        â†’
      </a>
    </div>
  </div>
  <div
    className="board__column"
  >
    <h3
      className="board__title"
    >
      Interviewing
    </h3>
  </div>
  <div
    className="board__column"
  >
    <h3
      className="board__title"
    >
      Hired
    </h3>
  </div>
</div>
`;
